contracts/extensions/FundingQueue.sol

```diff
@@ -0,0 +1,391 @@
+    require(proposal.state != ProposalState.Completed, "funding-queue-already-completed");
+    require(proposal.creator == msg.sender, "funding-queue-not-creator");
+    require(queue[_prevId] == _id, "funding-queue-bad-prev-id");
+
+    proposal.state = ProposalState.Cancelled;
+    proposal.lastUpdated = now;
+
+    queue[_prevId] = queue[_id];
+    delete queue[_id];
+
+    proposals[queue[_prevId]].lastUpdated = now;
+
+    emit ProposalCancelled(_id);
+  }
+
+  function stakeProposal(
+    uint256 _id,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    Proposal storage proposal = proposals[_id];
+
+    require(proposal.state == ProposalState.Inactive, "funding-queue-not-inactive");
+    require(proposal.creator == msg.sender, "funding-queue-not-creator");
+
+    proposal.state = ProposalState.Active;
+    proposal.domainTotalRep = checkReputation(_id, address(0x0), _key, _value, _branchMask, _siblings);
+
+    uint256 stake = wmul(proposal.domainTotalRep, STAKE_FRACTION);
+    colony.obligateStake(msg.sender, proposal.domainId, stake);
+
+    emit ProposalStaked(_id, proposal.domainTotalRep);
+  }
+
+  function backProposal(
+    uint256 _id,
+    uint256 _backing,
+    uint256 _currPrevId,
+    uint256 _newPrevId,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    Proposal storage proposal = proposals[_id];
+
+    require(proposal.state == ProposalState.Active, "funding-queue-proposal-not-active");
+    require(_id != _newPrevId, "funding-queue-cannot-insert-after-self"); // NOTE: this may be redundant
+
+    uint256 userRep = checkReputation(_id, msg.sender, _key, _value, _branchMask, _siblings);
+    require(_backing <= userRep, "funding-queue-insufficient-reputation");
+
+    // Update the user's reputation backing
+    uint256 prevBacking = supporters[_id][msg.sender];
+    if (_backing >= prevBacking) {
+      proposal.totalSupport = add(proposal.totalSupport, sub(_backing, prevBacking));
+    } else {
+      proposal.totalSupport = sub(proposal.totalSupport, sub(prevBacking, _backing));
+    }
+    supporters[_id][msg.sender] = _backing;
+
+    // Remove the proposal from its current position, if exists
+    require(queue[_currPrevId] == _id, "funding-queue-bad-prev-id");
+    queue[_currPrevId] = queue[_id];
+
+    // Insert into the right location
+    uint256 nextId = queue[_newPrevId];
+    if (queue[HEAD] == nextId && nextId != 0) {
+      pingProposal(nextId);
+    }
+    // Does this proposal have less than or equal support to the previous proposal?
+    require(
+      proposals[_newPrevId].totalSupport >= proposal.totalSupport,
+      "funding-queue-excess-support"
+    );
+    // Does this proposal have more than or equal support to the next proposal?
+    require(
+      nextId == HEAD || proposals[nextId].totalSupport < proposal.totalSupport,
```

Comment:
This `nextId==HEAD` really bugs me, which after staring at it for a while I think it's because this isn't a queue, it's a circular linked list. Can we characterise it more honestly in the code as such, or at least own up to that fact in the comments, and very explicitly call out this special case? Alternatively, we could make it more queue like with a `TAIL` that signifies the end of the queue.

Reply:
Mm that's fair.